target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/dsa.c

```diff
@@ -2294,47 +2299,43 @@ static int rtl83xx_port_mirror_add(struct dsa_switch *ds, int port,
 	if (group >= 4)
 		return -ENOSPC;
 
-	ctrl_reg = priv->r->mir_ctrl + group * 4;
-	dpm_reg = priv->r->mir_dpm + group * 4 * priv->port_width;
-	spm_reg = priv->r->mir_spm + group * 4 * priv->port_width;
-
 	pr_debug("Using group %d\n", group);
+
+	r = priv->r->get_mirror_config(priv, &config, group, mirror->to_local_port);
+	if (r < 0)
+		return r;
+
+	if (ingress)
+		pm_reg = config.spm;
+	else
+		pm_reg = config.dpm;
+
 	mutex_lock(&priv->reg_mutex);
 
-	if (priv->family_id == RTL8380_FAMILY_ID) {
-		/* Enable mirroring to port across VLANs (bit 11) */
-		sw_w32(1 << 11 | (mirror->to_local_port << 4) | 1, ctrl_reg);
-	} else {
-		/* Enable mirroring to destination port */
-		sw_w32((mirror->to_local_port << 4) | 1, ctrl_reg);
-	}
+	sw_w32(config.val, config.ctrl);
```

Comment:
To be safe, move the lock just before the group search loop.

Reply:
Yes, moved it and kept the unlock at the end of the function(s).